modules/utils/graph/Redif.hpp

```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+#include <vector>
+
+#include "3rdparty/blaze/Blaze.h"
+#include "modules/distance.hpp"
+
+namespace metric
+{
+	template<typename recType, class Metric = metric::Euclidian<recType>, typename Tv = double>
+	class Redif
+	{
+	public:
+		Redif(std::vector<std::vector<Tv>> trainData, size_t nNeighbors = 10, size_t nIter = 15);
+		~Redif() = default;
+		
+		blaze::DynamicMatrix<Tv> encode(blaze::DynamicMatrix<Tv> x, blaze::DynamicVector<Tv> & l_idx);
+		blaze::DynamicMatrix<Tv> decode(blaze::DynamicMatrix<Tv> xEncoded, blaze::DynamicMatrix<Tv> xTrainEncoded, blaze::DynamicVector<size_t> l_idx);

```

Comment:
very very wrong interface: do not use other inputs then x for encode and xEncode for decode methods. the internals must be hided in public methods.